### ISSUE TITLE
Multi-language support

### ISSUE THREAD

**Issue by hanbang-wang** [alecthomas/kingpin#172]
Is there a way to display different language according to system locale?

Though I can change the template in order to print help information in another language, I can't change some of the messages like `error: required flag xxx not provided...`.

So can you add multiple language support or allow us to change such error message.

**Comment by alecthomas** [comment]
I have no experience with multi-language support in Go. What would you suggest?

**Comment by alecthomas** [comment]
It looks like https://github.com/maximilien/i18n4go is quite popular.

**Comment by hanbang-wang** [comment]
Yep, you can try that.

But Go is pretty awkward in this territory, neither you can store files with compiled binary nor turn something variable to constant (in this case, string) during the compiling process. So you have to bring a json file or something else with you all the time for the multi-language support.

A sample solution would be, create a package like `gopkg.in/alecthomas/kingpin.v2/lang/en`, `github.com/alecthomas/kingpin/lang/zh_CN`, and take the idea from `database/sql`.

I'll write a similar approach.

**Comment by hanbang-wang** [comment]
OK, please go to [this commit of my fork](https://github.com/hanbang-wang/kingpin/commit/7d23abb3d4854ed33194129669967110145bd169#diff-a74dbf25619c2d0021f37b6e9f682226).

This would let Go only compile the languages needed.

PS:
- for simple POC, I used relative import.
- the implement of `type Lang` is not elegant.

**Comment by alecthomas** [comment]
I've given it an initial pass on the v3-unstable branch (where all active development is taking place). It seems reasonable to me, though the language is detected and initialised at application startup which is probably not desirable. The problem is there are some globals that need translation as well (eg. errors). I guess these could be inlined though.

**Comment by hanbang-wang** [comment]
Oh, I see you already changed your code. Ignore me than.

**Comment by alecthomas** [comment]
Sorry about that! I've been working on it all day, but only just saw your comment.

**Comment by hanbang-wang** [comment]
No, I didn't mean that, my apologies. My point is, such solution requires additional files, which is considerably a bad solution. One independent binary file with multi-language support and minimum file size is better.

**Comment by alecthomas** [comment]
Okay, yep. The language support is compiled in, so `go get` will work correctly.

**Comment by alecthomas** [comment]
If you're able to provide a translation, it would be fantastic.

**Comment by hanbang-wang** [comment]
Will do.
And I just went to see the source code, and found the function `ParseTranslationFileBytes` which did the thing, thanks.
And approximately release time of V3?

**Comment by alecthomas** [comment]
It is usable now, as `gopkg.in/alecthomas/kingpin.v3-unstable`. I think most of the major kinks have been worked out, but I'd say a month or two.

**Comment by hanbang-wang** [comment]
Looking forward to it.
I think the behavior should be like creating an application with, say, `kingpin.NewWithLang`, instead of parsing all the languages [here](https://github.com/alecthomas/kingpin/blob/8606e7e280c9bd28562395a8dd92d9e3a481e26f/i18n_init.go#L28-L29), and `SetLanguage` [here](https://github.com/alecthomas/kingpin/blob/8606e7e280c9bd28562395a8dd92d9e3a481e26f/i18n_init.go#L41).

**Comment by alecthomas** [comment]
I'm not against that, but there are two questions that arise:

1. Complication due to having to pass a reference to the translation function to every location that needs it.
2. What is the advantage of having language be that customisable?

**Comment by hanbang-wang** [comment]
The advantage would be reducing the binary file size. Go won't put unused global variable in the compiled binary file.
As to the first problem, I don't really understand, doesn't that `T` function need to lookup the language the user set everytime?

**Comment by alecthomas** [comment]
I don't think binary size is a huge concern, especially as each translation file is max 6k. Go binaries are typically in the realm of several MB.

My first point was related to the fact that if the language is tied to the `Application` struct (eg. `type Application struct { lang *Lang; ... }`) then in order to support simultaneous use of multiple languages, a reference to `lang *Lang` would need to be passed everywhere. However, if it is *not* tied to the struct, then there's no reason to set the language in the constructor.

So basically, I think the current approach is fine.

PS. I've added compression to the generator.
